core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -159,22 +158,26 @@ public void initialize(String name, Map<String, String> unresolved) {
     // build the final configuration and set up the catalog's auth
     Map<String, String> mergedProps = config.merge(props);
     Map<String, String> baseHeaders = configHeaders(mergedProps);
-    this.catalogAuth = new AuthSession(baseHeaders, null, null);
-    if (authResponse != null) {
-      this.catalogAuth = newSession(authResponse, startTimeMillis, catalogAuth);
-    } else if (initToken != null) {
-      this.catalogAuth = newSession(initToken, expiresInMs(mergedProps), catalogAuth);
-    }
-
+    this.client = clientBuilder.apply(mergedProps);
     this.sessions = newSessionCache(mergedProps);
     this.refreshAuthByDefault =
         PropertyUtil.propertyAsBoolean(
             mergedProps,
             CatalogProperties.AUTH_DEFAULT_REFRESH_ENABLED,
             CatalogProperties.AUTH_DEFAULT_REFRESH_ENABLED_DEFAULT);
-    this.client = clientBuilder.apply(mergedProps);
     this.paths = ResourcePaths.forCatalogProperties(mergedProps);
 
+    this.catalogAuth = new AuthSession(baseHeaders, null, null);
+    if (authResponse != null) {
+      this.catalogAuth =
+          AuthSession.sessionFromFetchedToken(
+              client, tokenRefreshExecutor(), authResponse, startTimeMillis, catalogAuth);
+    } else if (initToken != null) {
+      this.catalogAuth =
```

Comment:
I'm currently not 100% sure whether we already should check here if the `initToken` is expired and do a token exchange using the credential (similarly to how it's down further down in this class)

Reply:
Yes, this should check whether the expiration timestamp can be recovered and schedule based on that.

Reply:
moving the token expiration logic to `sessionFromToken` solves this